### PR TITLE
Organize LJ computation into an inline function

### PR DESCRIPTION
**Pull request by badisa**
* Because the order of operations is important due to how we handle
  exclusions combined with our fixed point tricks

Having been comparing this using NVProf, there is some fluctuation in values but haven't seen conclusive evidence that it has any impact on performance (which is what I would hope to see).

Original DHFR:
```
==244658== Profiling application: /home/ubuntu/.conda/envs/timemachine/bin/python /home/ubuntu/.conda/envs/timemachine/bin/pytest slow_tests/test_benchmark.py -k dhfr                                                                                                                                                      
==244658== Profiling result:                                                                                                                                                                                                                                                                                                
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name                                                                                                                                                                                                                                           
 GPU activities:   65.50%  273.17ms       297  919.77us  598.62us  1.4662ms  void k_nonbonded_unified<float, bool=0, bool=1, bool=0, bool=0>(int, double const *, double const *, double const *, double const *, double const *, double const *, double, double, double, int const *, unsigned int const *, __int64*, __int
64*, __int64*, __int64*)                                                                                                                                                                                                                                                                                                    
                    7.26%  30.280ms        44  688.19us  443.74us  1.1173ms  void k_find_blocks_with_ixns<float>(int, double const *, double const *, double const *, double const *, unsigned int*, int*, unsigned int*, unsigned int*, double)                                                                            
                    4.03%  16.825ms       300  56.082us  35.744us  89.664us  void gen_sequenced<curandStateXORWOW, double2, normal_args_double_st, __operator_&__(double2 curand_normal_scaled2_double<curandStateXORWOW>(curandStateXORWOW*, normal_args_double_st)), rng_config<curandStateXORWOW, curandOrdering=101>>(cu
randStateXORWOW*, double2*, unsigned long, unsigned long, normal_args_double_st)                                                                                                                                                                                                                                            
                    3.50%  14.609ms        17  859.32us  1.0240us  13.922ms  [CUDA memcpy HtoD]                                                                                                                                                                                                                             
                    3.38%  14.107ms       303  46.556us  22.272us  69.952us  void k_nonbonded_exclusions<float>(int, double const *, double const *, double const *, double const *, double const *, double const *, double, int const *, double const *, double, double, __int64*, __int64*, __int64*, __int64*)           
```

Functionalized DHFR
```
GPU activities:   65.64%  281.18ms       297  946.73us  592.48us  1.4664ms  void k_nonbonded_unified<float, bool=0, bool=1, bool=0, bool=0>(int, double const *, double const *, double const *, double const *, double const *, double const *, double, double, double, int const *, unsigned int const *, __int64*, __int
64*, __int64*, __int64*)                                                                                                                                                                                                                                                                                                    
                    7.30%  31.275ms        44  710.79us  436.77us  1.1058ms  void k_find_blocks_with_ixns<float>(int, double const *, double const *, double const *, double const *, unsigned int*, int*, unsigned int*, unsigned int*, double)                                                                            
                    4.05%  17.340ms       300  57.798us  35.456us  89.664us  void gen_sequenced<curandStateXORWOW, double2, normal_args_double_st, __operator_&__(double2 curand_normal_scaled2_double<curandStateXORWOW>(curandStateXORWOW*, normal_args_double_st)), rng_config<curandStateXORWOW, curandOrdering=101>>(cu
randStateXORWOW*, double2*, unsigned long, unsigned long, normal_args_double_st)                                                                                                                                                                                                                                            
                    3.41%  14.605ms        17  859.12us  1.0240us  13.917ms  [CUDA memcpy HtoD]                                                                                                                                                                                                                             
                    3.38%  14.459ms       303  47.718us  22.816us  70.080us  void k_nonbonded_exclusions<float>(int, double const *, double const *, double const *, double const *, double const *, double const *, double, int const *, double const *, double, double, __int64*, __int64*, __int64*, __int64)
```

This is just one of several runs I have had, which seems to show that there is quite a bit of variation between runs. @proteneer Would appreciate your thoughts on how to be more certain about this

### CODE REVIEW

**Comment by proteneer** [comment]
I'd run for longer, I expect there to be basically no difference in performance, `__forceinline__` should be a copy and paste code wise. Do you plan to do something similar for the electrostatics as well?

**Comment by badisa** [comment]
@proteneer Will run longer. I intend to do electrostatics as well, just wanted to start with a small piece to ensure I was doing the right thing.

**Comment by badisa** [comment]
Running the test_dhfr test with 2000 steps per batch seeing equivalent performance

Averages of Current implementation

k_nonbonded_exclusions<float>: 213.07 (213.31, 213.35, 212.55)
k_nonbonded_unified<float, bool=0, bool=1, bool=0, bool=0>: 3.7763 (3.78235, 3.78654, 3.76014)
k_nonbonded_unified<float, bool=0, bool=0, bool=0, bool=1>: 34.303 (34.508, 34.029, 34.372)
k_nonbonded_unified<float, bool=0, bool=1, bool=1, bool=0>: 2.6799 (2.6616, 2.6960, 2.6822)


Averages of inline function

k_nonbonded_exclusions<float>: 213.156 (212.78, 213.42, 213.27)
k_nonbonded_unified<float, bool=0, bool=1, bool=0, bool=0>: 3.7715 (3.76305, 3.78097, 3.77075)
k_nonbonded_unified<float, bool=0, bool=0, bool=0, bool=1>: 33.758 (33.722, 33.828, 33.724)
k_nonbonded_unified<float, bool=0, bool=1, bool=1, bool=0>: 2.6786 (2.6720, 2.6918, 2.6721)

**Comment by badisa** [comment]
Electrostatic changes shows the same performance on the DHFR test with 2000 steps per batch as with previous changes.

File: timemachine/cpp/src/kernels/k_nonbonded.cuh [hunk 1]
```diff
@@ -306,6 +306,92 @@ double __device__ __forceinline__ fix_nvidia_fmad(double a, double b, double c,
...
+    RealType sig_ij = sig_i + sig_j;
+
+    RealType sig_inv_dij = sig_ij*inv_dij;
+    RealType sig2_inv_d2ij = sig_inv_dij*sig_inv_dij;
+    RealType sig4_inv_d4ij = sig2_inv_d2ij*sig2_inv_d2ij;
+    RealType sig6_inv_d6ij = sig4_inv_d4ij*sig2_inv_d2ij;
+    RealType sig6_inv_d8ij = sig6_inv_d6ij*inv_d2ij;
+    RealType sig5_inv_d6ij = sig_ij*sig4_inv_d4ij*inv_d2ij;
+
+    // optimize this a little more
```
Comment (proteneer) [inline timemachine/cpp/src/kernels/k_nonbonded.cuh:375]: we can probably remove these comments now - I don't think they resulted in any speedsup

File: timemachine/cpp/src/kernels/k_nonbonded.cuh [hunk 2]
```diff
@@ -860,64 +915,42 @@ void __global__ k_nonbonded_exclusions(
...
-
-            RealType sig_grad = lj_scale*24*eps_ij*sig5_inv_d6ij*(2*sig6_inv_d6ij-1);
-            RealType eps_grad = lj_scale*4*(sig6_inv_d6ij-1)*sig6_inv_d6ij;
-
-            g_sigi += FLOAT_TO_FIXED_DU_DP<RealType, FIXED_EXPONENT_DU_DSIG>(-sig_grad);
-            g_sigj += FLOAT_TO_FIXED_DU_DP<RealType, FIXED_EXPONENT_DU_DSIG>(-sig_grad);
-
-            g_epsi += FLOAT_TO_FIXED_DU_DP<RealType, FIXED_EXPONENT_DU_DEPS>(-eps_grad*eps_j);
-            g_epsj += FLOAT_TO_FIXED_DU_DP<RealType, FIXED_EXPONENT_DU_DEPS>(-eps_grad*eps_i);
+            g_sigi -= FLOAT_TO_FIXED_DU_DP<RealType, FIXED_EXPONENT_DU_DSIG>(sig_grad);
```
Comment (proteneer) [inline timemachine/cpp/src/kernels/k_nonbonded.cuh:946]: I noticed that this changed the sign of -= and the term inside FLOAT_TO_FIXED_DU_DP(), was there a reason for this? In general I don't expect the two operations (-= and casting) to bitwise commute... should triple check or revert to old way.

Comment (badisa) [reply timemachine/cpp/src/kernels/k_nonbonded.cuh:946]: Will revert

**Review by proteneer — APPROVED** [review]
LGTM - feel free to merge when tests pass.

**Pull request merged**